### PR TITLE
planner: mark the both side operand of NAAJ & refuse partial column substitute in projection elimination of Apply de-correlation

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37032

Problem Summary:
two problems mainly:
1: in the projection elimination of Apply decorrelate rule, we try to substitute any condition used in Apply with raw projected expression under this projection, after which this projection can be eliminated. But we should make sure
that all columns mapped from the schema of projection OP must be substituted atomically (all succeed or fail).

2: Consider an EQ condition converted from NOT IN OP (NAAJ), in which we used to mark the right side column of this "NA" EQ as InOperand=true. In some cases like columns substitution as we talked about above, cases like:

  (a NA EQ b (marked InOperand=true ) ) where b will be substituted with like a constant string, so we got  (a NA EQ "example" ), this condition will not be considered as a NA EQ condition anymore and will be moved from other condition to equal condition of join OP in the logical rule of pddResolver .

### What is changed and how it works?
1: NAAJ is not just about one side, mark both as inOperand=true. so even the right has successfully been substituted with a constant string, we can also get (a (marked InOperand=true) NA EQ "example" ) as a NA EQ in the other condition.

2: make the column substitution atomically (all succeed or fail) in projection elimination of Apply decorrelation.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
make the both side operand of NAAJ & refuse partial column substitute in projection elimination
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
```
[2022-08-16T03:55:46.307Z] planner/core/expression_rewriter.go:940:6: ineffectual assignment to larg (ineffassign)

[2022-08-16T03:55:46.307Z] planner/core/logical_plans.go:396:58: should use make([]*expression.ScalarFunction, len(p.EqualConditions)) instead (S1019)

[2022-08-16T03:55:46.307Z] planner/core/rule_decorrelate.go:174:63: should use make([]expression.Expression, len(proj.Exprs)) instead (S1019)
```

@AilinKid Please fix linter.

File: planner/core/logical_plans.go [hunk 1]
```diff
@@ -375,21 +375,63 @@ func (p *LogicalJoin) GetPotentialPartitionKeys() (leftKeys, rightKeys []*proper
...
+// projection can't supply column8 anymore.
+//
+//    upper OP (depend on column8)   --> projection(constant "guo" --> column8)  --> lower layer OP
+//              |                                                       ^
+//              +-------------------------------------------------------+
+//
+//    upper OP (depend on column8)   --> lower layer OP
+//              |                             ^
+//              +-----------------------------+      // Fail: lower layer can't supply column8 anymore.
+//
```
Comment (hawkingrei) [inline planner/core/logical_plans.go:390]: I suggest you to use gofmt with go1.19. you will find some differences.

Comment (hawkingrei) [reply planner/core/logical_plans.go:390]: ditto

expression/util.go

File: planner/core/logical_plans.go [hunk 2]
```diff
@@ -375,21 +375,62 @@ func (p *LogicalJoin) GetPotentialPartitionKeys() (leftKeys, rightKeys []*proper
...
+//	upper OP (depend on column8)   --> lower layer OP
+//	          |                             ^
+//	          +-----------------------------+      // Fail: lower layer can't supply column8 anymore.
+func (p *LogicalJoin) columnSubstituteAll(schema *expression.Schema, exprs []expression.Expression) (hasFallback bool) {
+	// make a copy of exprs for convenience of substitution (may change/partially change the expr tree)
+	cpLeftConditions := p.LeftConditions.Clone()
+	cpRightConditions := p.RightConditions.Clone()
+	cpOtherConditions := p.OtherConditions.Clone()
+	cpEqualConditions := make([]*expression.ScalarFunction, len(p.EqualConditions))
```
Comment (winoros) [inline planner/core/logical_plans.go:395]: The `columnSubstituteImpl` doesn't modify the input arg in-place then return it as the return value.
So I think we don't need to clone it here.
Just making a slice and calling builtin `copy()` is enough

Comment (AilinKid) [reply planner/core/logical_plans.go:395]: make sense！columnSubstituteImpl only builds a new one from the bottom up when the lower args have changed.

**Review by winoros — APPROVED** [review]

File: cmd/explaintest/t/subquery.test [hunk 3]
```diff
@@ -20,3 +20,16 @@ drop table if exists t1, t2;
...
+create table stu(id int, name varchar(100));
+insert into stu values(1, null);
+create table exam(stu_id int, course varchar(100), grade int);
+insert into exam values(1, 'math', 100);
+set names utf8 collate utf8_general_ci;
+explain select * from stu where stu.name not in (select 'guo' from exam where exam.stu_id = stu.id);
+select * from stu where stu.name not in (select 'guo' from exam where exam.stu_id = stu.id);
+set names utf8mb4;
+explain select * from stu where stu.name not in (select 'guo' from exam where exam.stu_id = stu.id);
+select * from stu where stu.name not in (select 'guo' from exam where exam.stu_id = stu.id);
```
Comment (time-and-fate) [inline cmd/explaintest/t/subquery.test:35]: Use `explain format = 'brief'` to simplify the result.

**Review by time-and-fate — COMMENTED** [review]
Generally, I think "fallback" in the substitution-related code is not a good name.
Without the background of this issue, it's not easy to understand what does "fallback" mean, and what is fallback from what.
I think maybe "fail" or "incompatible" might be better.

File: expression/util.go [hunk 4]
```diff
@@ -384,44 +385,65 @@ func setExprColumnInOperand(expr Expression) Expression {
...
+	_, hasFallBack, resExpr := ColumnSubstituteImpl(expr, schema, newExprs, true)
+	return hasFallBack, resExpr
+}
+
 // ColumnSubstituteImpl tries to substitute column expr using newExprs,
 // the newFunctionInternal is only called if its child is substituted
-func ColumnSubstituteImpl(expr Expression, schema *Schema, newExprs []Expression) (bool, Expression) {
+// 1@bool means whether the expr has changed.
+// 2@bool means whether the expr should change (has the dependency in schema, while the corresponding expr has some compatibility), but finally fallback.
+// 3@Expression, the original expr or the changed expr, it depends on 1@bool
```
Comment (time-and-fate) [inline expression/util.go:406]: Is this format following any conventions or code styles?

File: planner/core/rule_decorrelate.go [hunk 5]
```diff
@@ -171,10 +171,15 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
 				// TODO: Actually, it can be optimized. We need to first push the projection down to the selection. And then the APPLY can be decorrelated.
 				goto NoOptimize
 			}
+			tempDecorrelatedProjExprs := make([]expression.Expression, len(proj.Exprs))
 			for i, expr := range proj.Exprs {
-				proj.Exprs[i] = expr.Decorrelate(outerPlan.Schema())
+				tempDecorrelatedProjExprs[i] = expr.Decorrelate(outerPlan.Schema())
 			}
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply planner/core/rule_decorrelate.go:177]: yes, make sense.

File: expression/util.go (hunk 4 above)
Comment (AilinKid) [reply expression/util.go:406]: not really, changed.

File: planner/core/rule_decorrelate.go [hunk 6]
```diff
@@ -171,10 +171,15 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
...
-			apply.columnSubstitute(proj.Schema(), proj.Exprs)
+			hasFallBack := apply.columnSubstituteAll(proj.Schema(), tempDecorrelatedProjExprs)
+			if hasFallBack {
+				goto NoOptimize
+			}
+			proj.Exprs = tempDecorrelatedProjExprs
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/rule_decorrelate.go]: You are right. I didn't notice that at first sight.

File: planner/core/rule_decorrelate.go [hunk 7]
```diff
@@ -171,10 +171,26 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
...
+			//	upper OP (depend on column8)   --> lower layer OP
+			//	          |                             ^
+			//	          +-----------------------------+      // Fail: lower layer can't supply column8 anymore.
+			hasFail := apply.columnSubstituteAll(proj.Schema(), proj.Exprs)
+			if hasFail {
+				goto NoOptimize
+			}
+			// step2: when it can be substituted all, we then just do the de-correlation.
 			for i, expr := range proj.Exprs {
 				proj.Exprs[i] = expr.Decorrelate(outerPlan.Schema())
```
Comment (time-and-fate) [inline planner/core/rule_decorrelate.go:192]: It seems like we also need to decorrelate the conditions in the `Apply` here. I guess currently failed cases are caused by this.

File: planner/core/rule_decorrelate.go [hunk 8]
```diff
@@ -171,10 +171,26 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan, opt *lo
...
 			}
```
Comment (time-and-fate) [inline planner/core/rule_decorrelate.go:193]: It seems like we also need to decorrelate the conditions in the `Apply` here. I guess currently failed cases are caused by this.

Comment (AilinKid) [reply planner/core/rule_decorrelate.go]: Ah, yes

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**